modules/service/src/main/scala/lucuma/itc/cache/EffectfulCache.scala

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) 2016-2023 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package lucuma.itc.cache
+
+import cats.effect.MonadCancelThrow
+import cats.syntax.all.*
+import io.chrisdavenport.keysemaphore.KeySemaphore
+import natchez.Trace
+import org.typelevel.log4cats.Logger
+
+/**
+ * Basic cache implementation that will invoke an effect to resolve an absent entry.
+ *
+ * Provides contention so that multiple concurrent requests for the same key are handled in a single
```

Comment:
is this necessary? if we have concurrent request, they may go more than once to redis, but maybe that's not so terrible. I'm not sure but perhaps the cost of a local lock is more than having some extra remote calls?

Reply:
The cost of a local memory lock is very low. The only place where I expect there to actually be contention is when explore is populating the configuration table, where it will make many calls to ITC with the same custom SED.

Explore is limiting itself to at most 4 concurrent requests, so without this ITC will call ODB (which in turn will call S3) 4 times instead of 1 whenever there's a cache miss.

This will only affect the retrieval of the custom SED, the actual computation is outside the contention.